src/conn/conn_compact.c

```diff
@@ -21,21 +21,269 @@ __compact_server_run_chk(WT_SESSION_IMPL *session)
+    WT_ERR(bm->size(bm, session, &compact_stat->start_size));
+    compact_stat->prev_compact_time = __wt_clock(session);
+
+    return (0);
+
+err:
+    __wt_free(session, compact_stat);
+
+    return (ret);
+}
+
+/*
+ * __wt_background_compact_end --
+ *     Fill resulting compact statistics in the background compact tracking list for a given file.
+ */
+int
+__wt_background_compact_end(WT_SESSION_IMPL *session)
+{
+    WT_BACKGROUND_COMPACT_STAT *compact_stat;
+    WT_BM *bm;
+    WT_CONNECTION_IMPL *conn;
+    wt_off_t bytes_recovered;
+    int64_t id;
+    const char *uri;
+
+    bm = S2BT(session)->bm;
+    id = S2BT(session)->id;
+    uri = bm->block->name;
+
+    compact_stat = __background_compact_get_stat(session, uri, id);
+    conn = S2C(session);
+
+    WT_RET(bm->size(bm, session, &compact_stat->end_size));
+    compact_stat->bytes_rewritten = bm->block->compact_bytes_rewritten;
+    bytes_recovered = compact_stat->start_size - compact_stat->end_size;
+
+    /*
+     * If the file failed to decrease in size, mark as an unsuccessful attempt. It's possible for
+     * compaction to do work (rewriting bytes) while other operations cause the file to increase in
+     * size.
+     */
+    if (bytes_recovered <= 0) {
+        compact_stat->consecutive_unsuccessful_attempts++;
+        compact_stat->prev_compact_success = false;
+    } else {
+        compact_stat->consecutive_unsuccessful_attempts = 0;
+        conn->background_compact.files_compacted++;
+        compact_stat->prev_compact_success = true;
+
+        /*
+         * Update the moving average of bytes rewritten across each file compact attempt. A
+         * weighting of 10% means that we are effectively considering the last 10 attempts in the
+         * average.
+         */
+        conn->background_compact.bytes_rewritten_ema =
+          (uint64_t)(0.1 * bm->block->compact_bytes_rewritten +
+            0.9 * conn->background_compact.bytes_rewritten_ema);
+    }
+
+    return (0);
+}
+
+/*
+ * __background_compact_list_cleanup --
+ *     Free all or any entry that has not been updated for more than a day in the background compact
+ *     tracking list.
+ */
+static void
+__background_compact_list_cleanup(
+  WT_SESSION_IMPL *session, WT_BACKGROUND_COMPACT_CLEANUP_TYPE cleanup_type)
+{
+    WT_BACKGROUND_COMPACT_STAT *compact_stat, *temp_compact_stat;
+    WT_CONNECTION_IMPL *conn;
+    uint64_t cur_time, i;
+
+    conn = S2C(session);
+    cur_time = __wt_clock(session);
+
+    for (i = 0; i < conn->hash_size; i++)
+        TAILQ_FOREACH_SAFE(
+          compact_stat, &conn->background_compact.compacthash[i], hashq, temp_compact_stat)
+        {
+            if (cleanup_type == BACKGROUND_CLEANUP_ALL_STAT ||
+              WT_CLOCKDIFF_SEC(cur_time, compact_stat->prev_compact_time) >
+                conn->background_compact.max_file_idle_time)
+                __background_compact_list_remove(session, compact_stat, i);
```

Comment:
Just out of curiosity, is everything working as expected when we remove an element? There are two for loops and one increases the index `i`. Is it still ok to increase it when an element is removed?

Reply:
Yes, the `TAILQ_FOREACH_SAFE` loop will exit when the list is empty, and there are `conn->hash_size` lists to check.